lib/ex_unit/lib/ex_unit/on_exit_handler.ex

```diff
@@ -7,24 +7,26 @@ defmodule ExUnit.OnExitHandler do
   end
 
   @spec register(pid) :: :ok
-  def register(pid) do
+  def register(pid) when is_pid(pid) do
     Agent.update(@name, &Map.put(&1, pid, []))
   end
 
-  @spec add(pid, term, fun) :: :ok | :error
-  def add(pid, ref, callback) do
+
+  @spec add(pid, term, (()-> term)) :: any
+  def add(pid, name_or_ref, callback) when is_pid(pid) and is_function(callback, 0) do
     Agent.get_and_update(@name, fn map ->
       if entries = Map.get(map, pid) do
-        entries = List.keystore(entries, ref, 0, {ref, callback})
+        entries = List.keystore(entries, name_or_ref, 0, {name_or_ref, callback})
         {:ok, Map.put(map, pid, entries)}
       else
         {:error, map}
       end
     end)
   end
 
-  @spec run(pid, non_neg_integer | :infinity) :: :ok | {Exception.kind, term, Exception.stacktrace}
-  def run(pid, timeout) do
+  @spec run(pid, timeout) :: :ok | {Exception.kind, term, Exception.stacktrace}
+  def run(pid, timeout)
+     when is_pid(pid) and (timeout == :infinity or (is_integer(timeout) and timeout >= 0)) do
```

Comment:
I don't think we should check the timeout, it feels very defensive. I think checking for pid and 0-arity are good because we only want to store/use valid data. If not a pid we will hide the error because nothing will be found in the map. However an invalid timeout will crash later. This is ok and will be clearer than a function clause error IMO.